hal/hal_remote/source/hal_socket_transmitter.cpp

```diff
@@ -0,0 +1,220 @@
+    bool print_port) {
+  if (!setup_connection_done) {
+    const hal_socket_transmitter::error_code res = setup_connection(true);
+    if (res != hal_socket_transmitter::success) {
+      last_error = res;
+      return res;
+    }
+    setup_connection_done = true;
+  }
+  if (listen() != 0) {
+    last_error = hal_socket_transmitter::listen_failed;
+    return last_error;
+  }
+  if (print_port) {
+    (void)printf("Listening on port %d\n", get_port());
+    (void)fflush(stdout);
+  }
+  if (accept() == -1) {
+    last_error = hal_socket_transmitter::accept_failed;
+    return last_error;
+  }
+  last_error = hal_socket_transmitter::success;
+  return last_error;
+}
+
+/// @brief make a connection to a server
+/// @return true if successful, otherwise will set last_error to errno
+hal_socket_transmitter::error_code hal_socket_transmitter::make_connection() {
+  if (!setup_connection_done) {
+    last_error = setup_connection(false);
+    if (last_error != hal_socket_transmitter::success) {
+      return last_error;
+    }
+    setup_connection_done = true;
+  }
+  last_error = connect();
+  return last_error;
+}
+
+bool hal_socket_transmitter::receive(void *data, uint32_t size) {
+  uint32_t data_to_read = size;
+  uint32_t offset = 0;
+
+  // repeatedly recv until `size` bytes is read.
+  do {
+    const int res = recv(fd_to_use, ((char *)data) + offset, data_to_read, 0);
+    // If recv returns 0, this indicates the connection has been dropped
+    // It's not an error as such but we are not able to continue
+    if (res == 0) {
+      is_connected = false;
+      fd_to_use = -1;
+      last_error = hal_socket_transmitter::connection_closed;
+      return false;
+    }
+    if (res == -1) {
+      last_error = hal_socket_transmitter::recv_error;
+      return false;
+    }
+    data_to_read = data_to_read - res;
+    offset += res;
+  } while (data_to_read);
+  last_error = hal_socket_transmitter::success;
+  return true;
+}
+
+/// @brief Send `size` bytes of `data` with an optional flush
+bool hal_socket_transmitter::send(const void *data, uint32_t size, bool flush) {
+  const int ret = ::send(fd_to_use, data, size, 0);
+  if (ret == -1) {
+    last_error = hal_socket_transmitter::send_error;
+    return false;
+  }
+  last_error = hal_socket_transmitter::success;
+  return true;
+}
+
+hal_socket_transmitter::error_code hal_socket_transmitter::connect() {
+  const int res = ::connect(sock, (struct sockaddr *)&server_address,
+                            sizeof(server_address));
+  if (res != -1) {
+    fd_to_use = sock;
+    is_connected = true;
+    return hal_socket_transmitter::success;
+  } else {
+    return hal_socket_transmitter::connect_failed;
+  }
+}
+
+hal_socket_transmitter::error_code hal_socket_transmitter::setup_connection(
+    bool server) {
+  // don't support port 0
+  if (port_requested == 0) {
+    if (debug_enabled()) {
+      (void)fprintf(stderr, "port 0 requested. This is disallowed\n");
```

Comment:
Given that port 0 is the default, this may not actually have been requested, would an error message along the lines of "The port to listen on must be specified" work?